examples/threshold_sig.rs

```diff
@@ -0,0 +1,203 @@
+// messages, the first message that has received `threshold + 1` signatures from validator nodes,
+// get's added to the `chat_log`.
+struct ChatNetwork {
+    pk_set: PublicKeySet,
+    nodes: Vec<Node>,
+    chat_log: ChatLog,
+    n_users: usize,
+}
+
+impl ChatNetwork {
+    // Creates a new network of nodes running our distributed chat protocol.
+    //
+    // # Arguments
+    //
+    // `n_nodes` - the number of validator/signing nodes in the network.
+    // `threshold` - our protocol requires a message to have `threshold + 1` validator signatures
+    // before it can be added to the `chat_log`.
+    fn new(n_nodes: usize, threshold: usize) -> Self {
+        let mut rng = rand::thread_rng();
+        let sk_set = SecretKeySet::random(threshold, &mut rng).unwrap();
+        let pk_set = sk_set.public_keys();
+
+        let nodes = (0..n_nodes)
+            .map(|id| {
+                let sk_share = sk_set.secret_key_share(id).unwrap();
+                Node::new(id, sk_share)
+            })
+            .collect();
+
+        ChatNetwork {
+            pk_set,
+            nodes,
+            chat_log: vec![],
+            n_users: 0,
+        }
+    }
+
+    fn create_user(&mut self) -> User {
+        let user_id = self.n_users;
+        let user = User::new(user_id);
+        self.n_users += 1;
+        user
+    }
+
+    fn get_node(&mut self, id: NodeId) -> &mut Node {
+        self.nodes
+            .get_mut(id)
+            .expect("No `Node` exists with that ID")
+    }
+
+    // Run a single round of the consensus algorithm. If consensus produced a new block, append
+    // that block the chat log. 
+    fn step(&mut self) {
+        if let Some(block) = self.run_consensus() {
+            self.chat_log.push(block);
+        }
+    }
+
+    // Our chat protocol's consensus algorithm. Produces a new block to be appended to the chat
+    // log. Our consensus uses threshold-signing to verify that a message has received enough
+    // signature shares (i.e. has been signed by `threshold + 1` nodes).
+    fn run_consensus(&self) -> Option<(UserId, Msg, Signature)> {
+        // Create a new `MsgDatabase` of every message that has been signed by a validator node.
+        let all_pending: MsgDatabase = self.nodes.iter()
+            .fold(BTreeMap::new(), |mut all_pending, node| {
+                for (user_id, signed_msgs) in node.pending.iter() {
+                    let mut user_msgs = all_pending
+                        .entry(*user_id)
+                        .or_insert(BTreeMap::new());
+                    for (msg, sigs) in signed_msgs.iter() {
+                        let sigs: Vec<NodeSignature> = sigs.iter().map(|sig| sig.clone()).collect();
+                        user_msgs
+                            .entry(msg.to_string())
+                            .or_insert(Vec::new())
+                            .extend(sigs);
```

Comment:
Maybe we should [verify](https://github.com/poanetwork/threshold_crypto/blob/3e6df11c4b20e8f2d315b48bae7dc14e903f0aa0/src/lib.rs#L185) the individual signature shares, too, and filter out the bad ones, so that a malicious node can't cause `combine_signatures` to fail.